### Q1
If $a$ is a constant, then
$$\kappa=\lim_{\delta\epsilon\rightarrow 0}\sup \frac{\lVert\delta f/f\rVert}{\lVert\delta b/b\rVert}.$$

If $b$ is a constant, then
$$\kappa=\lim_{\delta\epsilon\rightarrow 0}\sup \frac{\lVert\delta f/f\rVert}{\lVert\delta a/a\rVert}.$$

In [2]:
using LinearAlgebra

# Define the matrices
a = [10^10 0; 0 10^(-10)]
b = [10^10 0; 0 10^10]
c = [10^(-10) 0; 0 10^(-10)]
d = [1 2; 2 4]

# Compute the condition numbers
cond_a = cond(a, 2)
cond_b = cond(b, 2)
cond_c = cond(c, 2)
cond_d = cond(d, 2)


# Print the condition numbers
println("Condition number of matrix a: ", cond_a)
println("Condition number of matrix b: ", cond_b)
println("Condition number of matrix c: ", cond_c)
println("Condition number of matrix d: ", cond_d)

Condition number of matrix a: 1.0e20
Condition number of matrix b: 1.0
Condition number of matrix c: 1.0
Condition number of matrix d: 2.517588727560788e16


### Q2
Condition number for each matrix:
(a) $10^{20}$ (b) $1$ (c) $1$ (d) $2.52$.

So (a) and (d) are ill-conditioned, and (b) and (c) are well-conditioned.

### Q3

In [3]:
using LinearAlgebra

function gauss_jordan(A_in)
    A = copy(A_in)
    n = size(A,1)
    B = Matrix{Float64}(I, n, n)
    for i = 1:n
        for j = 1:n
            if j != i
                c_ji = A[j, i] / A[i, i]
                for k = 1 : i
                    B[j, k] -= B[i, k] * c_ji
                end 
                for k = i : n
                    A[j, k] -= A[i, k] * c_ji
                end
            end
        end
    end

    for i = 1:n
        for j in 1:n
            B[i, j] = B[i, j] / A[i, i]            
        end
    end
    return B
end

gauss_jordan (generic function with 1 method)

In [4]:
using Test
@testset "Gauss Jordan" begin
	n = 10
	A = randn(n, n)
	@test gauss_jordan(A) * A ≈ Matrix{Float64}(I, n, n)
end

Test Summary: | Pass  Total  Time
Gauss Jordan  |    1      1  0.9s


Test.DefaultTestSet("Gauss Jordan", Any[], 1, false, false, true, 1.683559672364168e9, 1.68355967323553e9)

#### Credits to Xuanzhao Gao